contracts/staking/StakingPool.sol

```diff
@@ -0,0 +1,594 @@
+        delete _users[_msgSender()];
+
+        bool noRewards = true;
+
+        for (uint256 i = 0; i < user.rewardAmounts.length; i++) {
+            if (user.rewardAmounts[i] > 0) {
+                noRewards = false;
+                //slither-disable-next-line calls-loop
+                _transferRewards(
+                    user.rewardAmounts[i],
+                    IERC20(_config.rewardTokens[i].tokens)
+                );
+            }
+        }
+        if (noRewards) {
+            emit NoRewards(_msgSender());
+        }
+    }
+
+    /**
+     * @notice Withdraw stake tokens when minimum pool conditions to begin are not met
+     */
+    function earlyWithdraw()
+        external
+        stakingPoolRequirementsUnmet
+        whenNotPaused
+    {
+        _withdrawWithoutRewards();
+    }
+
+    /**
+     * @notice Withdraw stake tokens when admin has enabled emergency mode
+     */
+    function emergencyWithdraw() external emergencyModeEnabled {
+        _withdrawWithoutRewards();
+    }
+
+    function initialize(
+        StakingPoolLib.Config calldata info,
+        bool paused,
+        uint32 rewardsTimestamp
+    ) external virtual initializer {
+        __RoleAccessControl_init();
+        __Context_init_unchained();
+        __Pausable_init();
+
+        //slither-disable-next-line timestamp
+        require(
+            info.epochStartTimestamp >= block.timestamp,
+            "StakingPool: start time"
+        );
+
+        _enforceUniqueRewardTokens(info.rewardTokens);
+        require(
+            address(info.stakeToken) != address(0),
+            "StakingPool: stake token defined"
+        );
+        //slither-disable-next-line timestamp
+        require(
+            rewardsTimestamp > info.epochStartTimestamp + info.epochDuration,
+            "StakingPool: init rewards"
+        );
+        require(info.treasury != address(0), "StakePool: nonzero treasury"); // TODO: are we checking if the treasury is whitelisted to that daoId
+        require(info.maxTotalPoolStake > 0, "StakePool: maxTotalPoolStake > 0");
+        require(info.epochDuration > 0, "StakePool: epochDuration > 0");
+        require(info.minimumContribution > 0, "StakePool: minimumContribution");
+
+        if (paused) {
+            _pause();
+        }
+
+        _rewardsAvailableTimestamp = rewardsTimestamp;
+        emit RewardsAvailableTimestamp(rewardsTimestamp);
+
+        _stakingPoolConfig = info;
+    }
+
+    function initializeRewardTokens(
+        address benefactor,
+        StakingPoolLib.Reward[] calldata rewards
+    ) external atLeastDaoMeepleRole(_stakingPoolConfig.daoId) {
+        _initializeRewardTokens(benefactor, rewards);
+    }
+
+    function enableEmergencyMode()
+        external
+        atLeastDaoAdminRole(_stakingPoolConfig.daoId)
+    {
+        _emergencyMode = true;
+        emit EmergencyMode(_msgSender());
+
+        _adminEmergencyRewardSweep();
```

Comment:
I think we should be able to enable the emergency mode without relying on `_adminEmergencyRewardSweep` being successful. This seems prone to exploitation by a malicious reward contract.

Reply:
I agree, unless there's a requirement that enabling the emergency mode demands the rewards are transferred to the Treasury.

question: what would happen if emergency mode is involved _before_ the rewards are fully funded?